bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs

```diff
@@ -0,0 +1,763 @@
+                .context(resources, "Unable to read kubeconfig")?;
+            let mgmt_config = Config::from_custom_kubeconfig(
+                mgmt_kubeconfig.to_owned(),
+                &KubeConfigOptions::default(),
+            )
+            .await
+            .context(resources, "Unable load kubeconfig")?;
+            let mgmt_k8s_client = kube::client::Client::try_from(mgmt_config)
+                .context(resources, "Unable create K8s client from kubeconfig")?;
+
+            create_vsphere_k8s_cluster(
+                &spec.configuration,
+                &mgmt_kubeconfig_path,
+                &mut resources,
+                &mut memo,
+            )
+            .await?;
+            // Retrieve the kubeconfig for the vSphere K8s workload cluster from the CAPI mgmt cluster
+            let k8s_secrets: Api<Secret> = Api::namespaced(mgmt_k8s_client, "eksa-system");
+            let kubeconfig_secret = k8s_secrets
+                .get(&format!("{}-kubeconfig", &spec.configuration.cluster.name))
+                .await
+                .context(
+                    resources,
+                    format!(
+                        "vSphere K8s cluster '{}' does not exist in CAPI mgmt cluster",
+                        &spec.configuration.cluster.name
+                    ),
+                )?;
+            let encoded_kubeconfig = kubeconfig_secret
+                .data
+                .context(resources, "Missing kubeconfig secret")?
+                .get("value")
+                .context(resources, "Missing base64-encoded kubeconfig secret value")?
+                .to_owned();
+            serde_json::to_string(&encoded_kubeconfig)
+                .context(
+                    resources,
+                    "Unable to serialize kubeconfig secret ByteString to String",
+                )?
+                .trim_matches('"')
+                .to_string()
+        } else {
+            spec.configuration
+                .cluster
+                .kubeconfig_base64
+                .to_owned()
+                .context(
+                    resources,
+                    "Kubeconfig for existing vSphere K8s cluster missing",
+                )?
+        };
+
+        // We are done, set our custom status to say so.
+        memo.current_status = "vSphere K8s cluster created".into();
+
+        client
+            .send_info(memo.clone())
+            .await
+            .context(resources, "Error sending final creation message")?;
+
+        Ok(CreatedVSphereK8sCluster {
+            cluster_name: spec.configuration.cluster.name,
+            endpoint: spec.configuration.cluster.control_plane_endpoint_ip,
+            encoded_kubeconfig,
+        })
+    }
+}
+
+async fn is_cluster_creation_required(
+    cluster_info: &VSphereK8sClusterInfo,
+    creation_policy: CreationPolicy,
+) -> ProviderResult<(bool, String)> {
+    let cluster_exists = does_cluster_exist(cluster_info).await?;
+    match creation_policy {
```

Comment:
This torturous function (my fault) is duplicated, I think. I wonder if it would make sense to take `cluster_exists` as input to this function and then the EKS and ECS agents could use the same function.

Reply:
I was wondering the same. I'll try refactoring this.